cirq/protocols/decompose.py

```diff
@@ -0,0 +1,234 @@
+import collections
+from typing import TYPE_CHECKING, Callable, Union, Any, Tuple, Iterable, \
+    TypeVar, List, Optional
+
+from typing_extensions import Protocol
+
+
+from cirq.type_workarounds import NotImplementedType
+
+if TYPE_CHECKING:
+    # pylint: disable=unused-import
+    import cirq
+
+
+TDefault = TypeVar('TDefault')
+
+TError = TypeVar('TError', Exception)
+
+RaiseTypeErrorIfNotProvided = ([],)
+
+
+def _value_error_describing_bad_operation(op: 'cirq.Operation') -> ValueError:
+    return ValueError(
+        "Operation doesn't satisfy the given `keep` "
+        "but can't be decomposed: {!r}".format(op))
+
+
+class SupportsDecompose(Protocol):
+    """An object that may be describable by a unitary matrix."""
+
+    def _decompose_(self) -> Union['cirq.OP_TREE', NotImplementedType]:
+        return NotImplemented
+
+
+class SupportsDecomposeWithQubits(Protocol):
+    """A gate that may be describable by a unitary matrix."""
+
+    def _decompose_(self, qubits: Tuple['cirq.QubitId', ...]
+                    ) -> Union['cirq.OP_TREE', NotImplementedType]:
+        return NotImplemented
+
+
+def _default_decomposer(op: 'cirq.Operation'
+                        ) -> Union['cirq.OP_TREE', NotImplementedType]:
+    method = getattr(op, '_decompose_', None)
+    return NotImplemented if op is None else method()
+
+
+def _intercept_decompose(
+        op: 'cirq.Operation',
+        decomposers: List[Callable[['cirq.Operation'], 'cirq.OP_TREE']]
+) -> 'cirq.OP_TREE':
+    for d in decomposers:
+        r = d(op)
+        if r is not NotImplemented:
+            return r
+    return NotImplemented
+
+
+def decompose(
+    val: Any,
+    *,
+    intercepting_decomposer: Callable[['cirq.Operation'],
+                                      'cirq.OP_TREE'] = None,
+    decomposer: Callable[['cirq.Operation'],
```

Comment:
I don't quite see the point of having the decomposer arg in decompose.  When would you pass this in instead of an intercepting_decomposer?

Reply:
Hmmm... think you're right. I introduced `decomposer` first so it felt essential to me, but all the use cases I've written ended up using `intercepting_decomposer`. You can even emulate `decomposer` by throwing an exception out of `intercepting_decomposer`.

I wish I could come up with a better name than `intercepting_decomposer`. Stuff like `first_decomposer` or `lead_decomposer` seems like it might be confused with "at the top level you apply this one, but not when recursing".

Reply:
Done.